compiler/src/dmd/dmangle.d

```diff
@@ -231,235 +240,376 @@ unittest
+        visitType(t);
         mangleSymbol(t.sym);
     }
 
-    override void visit(TypeClass t)
+    void visitTypeClass(TypeClass t)
     {
         //printf("TypeClass.toDecoBuffer('%s' mod=%x) = '%s'\n", t.toChars(), mod, name);
-        visit(cast(Type)t);
+        visitType(t);
         mangleSymbol(t.sym);
     }
 
-    override void visit(TypeTuple t)
+    void visitTypeTuple(TypeTuple t)
     {
         //printf("TypeTuple.toDecoBuffer() t = %p, %s\n", t, t.toChars());
-        visit(cast(Type)t);
+        visitType(t);
         Parameter._foreach(t.arguments, (idx, param) {
-                mangleParameter(param);
+                mangleParameter(param, buf, backref);
                 return 0;
         });
         buf.writeByte('Z');
     }
 
-    override void visit(TypeNull t)
+    void visitTypeNull(TypeNull t)
     {
-        visit(cast(Type)t);
+        visitType(t);
     }
 
-    override void visit(TypeNoreturn t)
+    void visitTypeNoreturn(TypeNoreturn t)
     {
         buf.writestring("Nn");
     }
 
+    if (modMask != t.mod)
+    {
+        MODtoDecoBuffer(buf, t.mod);
+    }
+    if (backref.addRefToType(buf, t))
+        return;
+
+    switch (t.ty)
+    {
+        case Tvoid:
+        case Tint8:
+        case Tuns8:
+        case Tint16:
+        case Tuns16:
+        case Tint32:
+        case Tuns32:
+        case Tint64:
+
+        case Tuns64:
+        case Tfloat32:
+        case Tfloat64:
+        case Tfloat80:
+        case Timaginary32:
+        case Timaginary64:
+        case Timaginary80:
+        case Tcomplex32:
+        case Tcomplex64:
+        case Tcomplex80:
+
+        case Tbool:
+        case Tchar:
+        case Twchar:
+        case Tdchar:
+
+        case Tint128:
+        case Tuns128:    visitType(t); break;
+
+        case Tpointer:
+        case Treference:
+        case Tdelegate:
+        case Tslice:     visitTypeNext      (cast(TypeNext)t);      break;
+
+        case Tarray:     visitTypeDArray    (t.isTypeDArray());     break;
+        case Tsarray:    visitTypeSArray    (t.isTypeSArray());     break;
+        case Taarray:    visitTypeAArray    (t.isTypeAArray());     break;
+        case Tfunction:  visitTypeFunction  (t.isTypeFunction());   break;
+        case Tident:     visitTypeIdentifier(t.isTypeIdentifier()); break;
+        case Tclass:     visitTypeClass     (t.isTypeClass());      break;
+        case Tstruct:    visitTypeStruct    (t.isTypeStruct());     break;
+        case Tenum:      visitTypeEnum      (t.isTypeEnum());       break;
+        case Ttuple:     visitTypeTuple     (t.isTypeTuple());      break;
+        case Tnull:      visitTypeNull      (t.isTypeNull());       break;
+        case Tvector:    visitTypeVector    (t.isTypeVector());     break;
+        case Tnoreturn:  visitTypeNoreturn  (t.isTypeNoreturn);     break;
+
+        case Terror:
+            break;      // ignore errors
+
+        default:
+            assert(0);
```

Comment:
This was reverted because it caused a regression.

https://github.com/dlang/dmd/pull/14477

Reply:
It would help if you said what the regression was.

Reply:
I did in the introducing pr
https://github.com/dlang/dmd/pull/14468#discussion_r978712868